### ISSUE TITLE
[request] (really) dense point cloud

### ISSUE THREAD

**Issue by andrea-bistacchi** [alicevision/meshroom#1670]
Hello, a lot of people working in photogrammetry are more interested in a dense point cloud than in a textured surface. However **Meshroom yields very poor dense clouds**, that are generally between **one and two order of magnitude less dense** than those obtained with Agisoft or CaptureReality, and also VisualSFM point clouds are one order of magnitude denser.

We suspect that this is mainly a problem of setting the parameters right in the MVS step, and then exporting the full-resolution point cloud instead of a filtered one. Many issues have been opened dealing with this problem (#1462, #1302, #1203,  #1132, #888, #885, #803, #690,  #382, #266, etc.), but no real solution has been proposed.

Please help us, otherwise Meshroom will be useless for us (and for all people interested in point clouds), and this would really be a pity since it is a great project, really!

**Comment by andrea-bistacchi** [comment]
Hello, thanks for answering but this is what we are exporting, and it is not dense when compared to the output by other codes.

If you like we can send some example, but I guess we cannot attach them here (we need some GB).

I guess Meshroom works in principle as the other codes, so we need some advice on the "settings in other nodes to increase the point density".

Thanks very much!

**Comment by fabiencastan** [comment]
Yes, it would be great if you could share some concrete examples with the number of points you get in the different solutions, so we can reproduce it and provide precise recommendations.

My email to send the datasets: fabien.castan@technicolor.com (you can use fromsmash.com or wetransfer.com).

**Comment by andrea-bistacchi** [comment]
Hi, one of my students will send all details tomorrow. We are really struggling with all parameters and we cannot find a solution.

If you like we can send the whole projects with photos etc., but in this case we need an email address.

Thanks very much!!!

**Comment by stefano-casiraghi** [comment]
Hi, I'm the student who works with Andrea. We are mainly interested in the 3D point cloud of natural outcrops. Reading through the user's guide I found a few parameters to work with, for instance  in the Feature Extraction node I set the Describer Quality and the Describer Density to High,  

![Feature Extraction](https://user-images.githubusercontent.com/104076795/164237395-ed7b7ec2-48fd-4a2c-bbbb-e090e0a53cab.png)

and in the Meshing node, I maximize the  Max Input Points, Max Points, and Max points per voxel options.

![meshing](https://user-images.githubusercontent.com/104076795/164238557-683cc7b3-2d1e-4e0a-8c2b-7f72b8931dfc.png)

The resulting point cloud is characterized by a very poor point density (490,000 points), and the outcrop surface is noisier if compared to a point cloud coming from a standard Agisoft Photoscan workflow (18,000,000 points). 

Meshroom:

![Point cloud](https://user-images.githubusercontent.com/104076795/164249153-6758d605-5ac6-4bba-9f74-30ba5b8a8aec.png)

Agisoft (Visualized in Cloud Compare):

![agisoft_point_cloud](https://user-images.githubusercontent.com/104076795/164248389-aaf58136-e69b-463f-a478-0feef5d96f58.png)
![agidoft_punti](https://user-images.githubusercontent.com/104076795/164248415-828de76d-3f24-4098-a96c-66bd8efd26e5.png)

Any advice will be highly appreciated, 

Thanks very much!

**Comment by natowi** [comment]
I compared the results of Agisoft and Meshroom. Yes, Agisoft has a really Dense Point Cloud by default, but Meshroom achieves a Mesh with a similar and sometimes even better quality:

![grafik](https://user-images.githubusercontent.com/8138537/164433434-96b8fdb1-b4c3-46b9-b103-8872a9d4689c.png)

From my experience, Meshroom focuses on robust reconstructions and discards many points that are not robust enough by default:
![grafik](https://user-images.githubusercontent.com/8138537/164434126-b00c9b8d-c259-47b8-855b-63bde82e19e9.png)
![grafik](https://user-images.githubusercontent.com/8138537/164434222-86d2063e-1316-4f91-815b-15b7c1f2a3e0.png)
Areas that were visible in the sparse StructureFromMotion point cloud and DepthMaps don´t make it into the final reconstruction. 

From the Meshing log it looks like we internally have a point cloud density similar to Agisoft, but it is filtered down later:
![grafik](https://user-images.githubusercontent.com/8138537/164441879-c8eea64b-07b1-4caf-8000-fa44c8b1a8b7.png)


By tweaking the parameters I was able to increase the point cloud density for three images from the monstree dataset from 32K points (default) to 450K points - but the Agisoft point cloud is far denser still (4.8M points)

Here are the parameters I adjusted:
Meshing: max input points, max points, max points per voxel - to maximum value and Add Landmarks To The Dense Point Cloud and Colorize Output enabled, Densify enabled, Densify front/back 5
DepthMapFilter: relax the Min Consistent Cameras and Min Consistent Cameras Bad Similarity to 2 and 3 respectively (completely removing this node can increase the point cloud density with the downside of having noise)
DepthMap: Downscale 1
Enable additional describers and adjust the quality and density (depends on your dataset)

**Comment by andrea-bistacchi** [comment]
Hello, and thanks!

I totally agree on mesh quality. In fact quite often in Agisoft you need to clean the dense point cloud before meshing.

However for some application we do not us the mesh but the **point cloud directly**, and this can be very dense and still accurate if the photo dataset is very redundant (i.e. thousands of photo, with 6 to 10 photos looking at each portion of the model).

We'll try to work on the parameters you highlighted.

Thanks!

**Comment by fabiencastan** [comment]
DepthMap: Downscale=1 is the main one.
Meshing: PixSizeMarginInitCoef=1, PixSizeMarginFinalCoef=1, MinStep=1,

**Comment by prwurech** [comment]
Hello, thanks for the suggestions, but I also fail creating a dense point cloud. The settings suggested above have no result for me.

Pipeline 1
![Image](https://github.com/user-attachments/assets/d4815cc8-08d2-4382-9062-96d09ffdcf42)

Pipeline 2
![Image](https://github.com/user-attachments/assets/d90e71ef-85ba-4d9e-a240-1c07be0c6cf6)

Pipeline 2 is 25% faster than Pipeline 1, almost no difference in point density.
Both Meshroom Pipelines 1+2 create 2.5mio points, whereas Agisoft creates 50mio points in less time.

Has anyone other settings to create dense points clouds?
Best.

**Comment by natowi** [comment]
@prwurech If you just need a point cloud, you could convert the mesh: https://www.youtube.com/watch?v=T4A3rH7Uz3c

**Comment by prwurech** [comment]
Thanks @natowi for mentioning the sampling trick in CC, however it only adds geometry without adding detail. I seek to obtain maximum detail with a dense point cloud.

Interestingly, by enabling "Save Raw Dense Point Cloud" in the "Meshing" step, I get a 25% denser cloud (saved in .abc format) in comparison to the ConvertSfMFormat (saved in .ply format). Yet there seems to be no option for exporting the .abc file from the viewer.

To be fair, as its name says, Meshroom might be optimized for meshes, and generating dense point clouds is not really a feature. tbc
Best.

**Comment by andrea-bistacchi** [comment]
Hello, we tried a lot a couple of years ago, then we gave up. No way to get a properly dense point cloud from Meshlab.

It's a pity because - I regret to say - this leaves us without an open-source solution that could be competitive  with commercial software.
